# (연습) 모듈

**참고**

먼저 [모듈](https://codingalzi.github.io/code-workout-python/modules/) 내용을 공부한 다음에 연습문제를 풀어볼 것을 권장한다.

## `random` 모듈

### 연습문제

**문제 1**

1에서 99까지의 정수 중에서 3의 배수를 무작위로 세 개 생성하는 코드를 작성하라.

힌트: `random.randrange()` 함수와 `for` 반복문 활용.

답:

`random.randrange(a, b, c)` 형식으로 함수 호출이 되면
`a`에서 `b` 이전 까지의 구간에서 값을 무작위로 선택할 때 아래에 언급된 값 중에서 하나 선택한다.

    a, a+3, a+6, a+9, ...

따라서 1에서 99사이에서 3의 배수만을 대상으로 하려면
3의 배수만 선택 대상으로 지정되도록 셋째 인자를 3으로 지정한다.
단, 구간의 시작또한 3의 배수로 지정해야 3의 배수만을 대상으로 한다.
1보다 같거나 큰 정수 중에서 3이 가장 작은 3의 배수이기에
구간의 시작을 3으로 해야 한다.

아래 코드를 실행하면 1에서 99까지의 정수 중에서 3의 배수를
무작위로 세 개 생성한다.

In [ ]:
import random

for _ in range(3):
    print(random.randrange(3, 100, 3), end=', ')

15, 90, 48, 

이전 코드에서 `for _ in range(3)`을 사용하는 이유는
3의 배수를 선택하여 출력하는 명령문을 무조건 세 번 실행하기 위해서이다.
참고로 밑줄 기호 `_`는 `for` 반복문에 사용되는 변수의 
이름이 중요하지 않은 경우에 대신 사용한다.

실제로 `_` 대신에 `i`, `count` 등을 사용하더라도
`for` 반복문의 본문에 해당 변수가 사용되지 않기에 
변수의 이름이 전혀 중요하지 않다.
또한 `print()` 함수에서 `end=', '` 키워드 인자를 사용해서
무작위로 선택된 세 개의 값이 한 줄에 출력되도록 하였다.

**문제 2**

아래 `students` 리스트에 담긴 학생 이름을 `random.shuffle()` 함수로 무작위로 섞은 뒤,
`for` 반복문을 이용하여 발표 순서를 번호와 함께 출력하는 코드를 작성하라.

```python
students = ["김민준", "이서연", "박지호", "최수아", "정도윤"]
```

출력 예시:
```
발표 순서:
1번: 최수아
2번: 박지호
3번: 김민준
4번: 정도윤
5번: 이서연
```

답:

`random.shuffle()`은 원본 리스트를 직접 변경하므로 반환값 없이 호출한 뒤
`for` 반복문으로 인덱스와 함께 출력한다.

In [ ]:
import random

students = ["김민준", "이서연", "박지호", "최수아", "정도윤"]

random.shuffle(students)

print("발표 순서:")
for i in range(len(students)):
    print(str(i + 1) + "번: " + students[i])

발표 순서:
1번: 이서연
2번: 최수아
3번: 정도윤
4번: 김민준
5번: 박지호


**문제 3**

`random` 모듈의 `uniform()` 함수가 `random.uniform(a, b)` 형식으로 호출되면
구간 `a` 이상, `b` 미만의 구간에서 무작위로 하나의 부동소수점을
선택해서 반환한다.
실제로 `random.uniform(a, b)`는 다음과같이 정의되었다.

```python
random.uniform(a, b) = a + (b - a) * random.random()
```

`uniform`이란 단어의 의미는 지정된 구간에서 균등한 확률로 부동소수점을 선택한다는 의미이다.
이를 확인하기 위해 `random.uniform(0, 1)`을 10만 번 호출했을 때
0.1 보다 작은 값이 나온 경우가 10% 정도임을 확인하는 코드를 작성하라.

힌트: `for` 반복문, `range()` 함수, `random.uniform()` 함수 활용

답:

`random.uniform(0, 1)`를 10만 번 호출해서 반환값이 0.1보다 작은 경우의 비율을 계산하는 코드는 다음과 같다.

In [ ]:
import random

total = 100000
count = 0

for _ in range(total):
    if random.uniform(0, 1) < 0.1:
        count += 1
        
print(count / total)
    

0.10084


**문제 4**

베스킨라빈스 31 게임은 두 사람이 참여하며 마지막에 31을 부른 사람이 지는 게임으로 규칙은 다음과 같다.

* 참여자들은 번갈아가며 1부터 31까지의 수를 순서대로 부른다.
* 한번에 1~3개의 수를 부를 수 있다.

예를 들어, playerA가 1, 2, 3을 부르면, playerB는 4 또는 4, 5 또는 4, 5, 6를 부를 수 있다.

두 대의 컴퓨터 playerA와 playerB가 진행하는 베스킨라벤스 31 게임을 다음과 같이 구현할 수 있다.

아래 코드의 `beskinRobbins31()` 함수를 실행하면 
playerA와 playerB 모두 승리 전략이 없이 무작위로 숫자를 부르고 31을 부르는 컴퓨터가 결정되면
해당 컴퓨터 이름을 반환한다.
원래 각 컴퓨터가 부르는 숫자를 화면에 출력해야 게임이 진행되는 과정을 볼 수 있지만 생략하고 최종 결과만 보여주도록 하였다.

In [ ]:
import random

def beskinRobbins31():
    final = 31 # 언급되면 지는 숫자
    call = 0   # 마지막으로 언급된 숫자 기억
    player = 0 # 숫자를 불러야 하는 게임 참여자. 0은 playerA, 1은 playerB를 가리킴.

    while call < final:

        if player == 1:
            player_name = 'playerB'
        else:
            player_name = 'playerA'

        number = random.randint(1, 3)

        count = 0
        while count < number:
            call += 1
            if call == final:
                return player_name # 게임에 진 사용자 반환
            count += 1

        player = (player + 1) % 2

In [ ]:
print(beskinRobbins31() + "가 졌습니다.")

playerB가 졌습니다.


In [ ]:
print(beskinRobbins31() + "가 졌습니다.")

playerA가 졌습니다.


`beskingRobbins31()` 게임을 십만 번 진행했을 때 각 컴퓨터의 승률를 확인하는 코드를 작성해서 두 참여자 각각 50%의 승률을 가짐을 보여라.

답:

아래 코드는 베스킨라벤스 게임을 십만 번 진행했을 때 playerA의 승률이 대략적으로 0.5, 즉 50%임을 보인다.

In [ ]:
# playerA가 이긴 경우의 수
count = 0

# 십만 번 게임 진행
total_play = 100_000

num = 0
while num < total_play:
    if beskinRobbins31() == 'playerB':
        count += 1
    num += 1

print(count / total_play)

0.50199


## `statistics` 모듈

### 연습문제 

**문제 1**

아래 `classA`와 `classB`는 두 반의 시험 점수를 담은 리스트이다.

In [ ]:
classA = [82, 76, 91, 88, 95, 88, 73, 88, 90, 84]
classB = [90, 92, 45, 88, 91, 93, 89, 94, 90, 88]

(1) `statistics` 모듈을 `stats` 별칭으로 불러온 뒤, 두 반 각각의 평균값, 중앙값, 최빈값, 표준편차를 출력하라.

답:

`statistics` 모듈의 `mean()`, `median()`, `mode()`, `stdev()` 함수를 이용하여
두 반 각각의 통계량을 출력한다.
`classA`에서 88이 세 번 등장하므로 최빈값은 88이다.
`classB`에서 90이 리스트에서 먼저 등장하므로 최빈값은 90이다.

In [ ]:
import statistics as stats

print("=== classA ===")
print("평균값:", stats.mean(classA))
print("중앙값:", stats.median(classA))
print("최빈값:", stats.mode(classA))
print("표준편차:", stats.stdev(classA))

print()

print("=== classB ===")
print("평균값:", stats.mean(classB))
print("중앙값:", stats.median(classB))
print("최빈값:", stats.mode(classB))
print("표준편차:", stats.stdev(classB))

=== classA ===
평균값: 85.5
중앙값: 88.0
최빈값: 88
표준편차: 6.83536555146996

=== classB ===
평균값: 86
중앙값: 90.0
최빈값: 90
표준편차: 14.544949486180952


(2) 두 반의 평균값을 비교하여 평균이 더 높은 반의 이름(`"classA"` 또는 `"classB"`)을 출력하라. 단, `if` 조건문을 활용하라.

답:

`stats.mean()` 함수의 반환값을 변수에 저장한 뒤 `if` 조건문으로 비교한다.
`classA`의 평균은 85.5, `classB`의 평균은 86.0이므로 `"classB"`가 출력된다.

In [ ]:
mean_A = stats.mean(classA)
mean_B = stats.mean(classB)

if mean_A > mean_B:
    print("평균이 더 높은 반: classA")
else:
    print("평균이 더 높은 반: classB")

평균이 더 높은 반: classB


(3) 표준편차가 더 작은 반이 성적이 더 고르다고 할 수 있다. 두 반 중 성적이 더 고른 반의 이름을 출력하라.

답:

표준편차가 작을수록 점수들이 평균 근처에 모여 있다는 뜻이다.
`classA`의 표준편차는 약 6.85인 반면에 `classB`의 표준편차는 약 14.55이다.
`classB`에 45점이라는 이상치가 포함되어 있어 표준편차가 크다.
따라서 성적이 더 고른 반은 `classA`이다.

In [ ]:
stdev_A = stats.stdev(classA)
stdev_B = stats.stdev(classB)

if stdev_A < stdev_B:
    print("성적이 더 고른 반: classA")
else:
    print("성적이 더 고른 반: classB")

성적이 더 고른 반: classA


(4) 두 리스트를 하나로 합친 `total` 리스트를 만든 뒤, 전체 20명의 평균값과 중앙값을 구하라. 그리고 `for` 반복문을 활용하여 평균 이상인 점수의 개수를 구하라.

답:

두 리스트를 `+` 연산으로 합치면 20개의 점수를 담은 `total` 리스트가 만들어진다.
전체 평균값은 85.75이고 중앙값은 88.5이다.
`for` 반복문으로 각 점수가 평균 이상인지 확인하면 총 15개이다.
중앙값이 평균값보다 큰 이유는 `classB`의 45점이 평균을 아래로 끌어내리기 때문이다.

In [ ]:
total = classA + classB

total_mean = stats.mean(total)
total_median = stats.median(total)

print("전체 평균값:", total_mean)
print("전체 중앙값:", total_median)

count = 0
for score in total:
    if score >= total_mean:
        count += 1

print("평균 이상인 점수의 개수:", count)

전체 평균값: 85.75
전체 중앙값: 88.5
평균 이상인 점수의 개수: 15


**문제 2**

`random` 모듈과 `statistics` 모듈을 함께 활용하는 문제이다.

In [ ]:
import random
import statistics as stats

(1) `random.seed(42)`를 설정한 뒤, `random.randint(50, 100)`을 15번 호출하여 무작위 점수 15개를 담은 리스트 `random_scores`를 만들어라.

힌트: `for` 반복문과 리스트의 이어붙이기 연산을 활용하라.

답:

`random.seed(42)`를 먼저 호출하여 시드를 고정한 다음,
`for` 반복문을 15번 반복하면서
`random.randint(50, 100)`의 반환값으로 생성된 값을
지정된 리스트에 리스트 이어붙이기 연산을 이용하여 추가한다.

아래 코드는 
시드 값이 42로 고정되어 있으므로 매번 실행해도 동일한 15개의 점수가 생성된다.

In [ ]:
random.seed(42)

random_scores = []
for _ in range(15):
    random_scores += [random.randint(50, 100)]

print(random_scores)

[90, 57, 51, 97, 67, 65, 64, 58, 97, 56, 93, 97, 84, 55, 87]


(2) `random_scores`의 평균값, 중앙값, 최빈값, 표준편차를 출력하라.

답:

In [ ]:
print("평균값:", stats.mean(random_scores))
print("중앙값:", stats.median(random_scores))
print("최빈값:", stats.mode(random_scores))
print("표준편차:", stats.stdev(random_scores))

평균값: 74.53333333333333
중앙값: 67
최빈값: 97
표준편차: 17.844033283132035


(3) `for` 반복문을 활용하여 `random_scores`에서 평균 미만인 점수만 모은 리스트 `below_mean`을 만들고 출력하라.

답:

먼저 `stats.mean()`으로 평균값을 구한 뒤,
`for` 반복문으로 각 점수가 평균 미만인지 `if` 조건문으로 확인한다.
조건을 만족하는 점수만 `below_mean` 리스트에 추가한다.

In [ ]:
avg = stats.mean(random_scores)
below_mean = []

for score in random_scores:
    if score < avg:
        below_mean += [score]

print("평균 미만인 점수:", below_mean)
print("개수:", len(below_mean))

평균 미만인 점수: [57, 51, 67, 65, 64, 58, 56, 55]
개수: 8


(4) 모든 학생에게 5점의 보너스 점수를 주되, 100점을 초과할 수 없도록 하여 `adjusted_scores` 리스트를 만들어라. 그리고 보정 전후의 평균값과 표준편차를 비교하라. 단, `for` 반복문과 `if` 조건문을 활용하라.

답:

`for` 반복문으로 각 점수에 5점을 더한 뒤,
`if` 조건문으로 100을 초과하면 100으로 제한한다.
보너스를 주면 평균값은 올라가지만,
100점 상한이 있으므로 높은 점수는 덜 올라가서
표준편차가 약간 줄어들 수 있다.

In [ ]:
adjusted_scores = []

for score in random_scores:
    new_score = score + 5
    if new_score > 100:
        new_score = 100
    adjusted_scores += [new_score]

print("보정 전 평균:", stats.mean(random_scores))
print("보정 후 평균:", stats.mean(adjusted_scores))
print()
print("보정 전 표준편차:", stats.stdev(random_scores))
print("보정 후 표준편차:", stats.stdev(adjusted_scores))

보정 전 평균: 74.53333333333333
보정 후 평균: 79.13333333333334

보정 전 표준편차: 17.844033283132035
보정 후 표준편차: 17.315833657034688
